core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,201 @@
+
+  /**
+   * Constructor for a lock cache.
+   *
+   * @param defaultLoader specify a function to generate a value based on a key
+   * @param initialSize initial size of the cache
+   * @param maxSize maximum size of the cache
+   * @param concurrencyLevel concurrency level of the cache
+   */
+  public LockCache(@Nullable Function<? super K, ? extends V> defaultLoader, int initialSize,
+      int maxSize, int concurrencyLevel) {
+    mDefaultLoader = defaultLoader;
+
+    mHardLimit = maxSize;
+    mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
+    mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
+    mIterator = mCache.entrySet().iterator();
+    mEvictLock = new ReentrantLock();
+  }
+
+  private void evictIfOverLimit() {
+    long numToEvict = mCache.size() - mSoftLimit;
+
+    if (numToEvict <= 0) {
+      return;
+    }
+    // this will block if every lock has a reference on them.
+    if (mEvictLock.tryLock()) {
+      try {
+        // This thread is the evictor
+        while (numToEvict > 0) {
+          if (!mIterator.hasNext()) {
+            mIterator = mCache.entrySet().iterator();
+          }
+          Map.Entry<K, ValNode<V>> candidateMapEntry = mIterator.next();
+          ValNode<V> candidate = candidateMapEntry.getValue();
+          if (candidate.mIsAccessed) {
+            candidate.mIsAccessed = false;
+          } else {
+            if ((!candidate.mIsNew) && candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
+              // the value object can be evicted, at the same time we make refCount minValue
+              mIterator.remove();
+              numToEvict--;
+            }
+          }
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * get the value from the cache.
+   *
+   * @param key the key to look up the cache
+   * @return the value contained in the cache, if it is already in cache,
+   * otherwise generate an entry based on the loader
+   */
+  public ValNode<V> get(final K key) {
+    Preconditions.checkNotNull(key, "key can not be null");
+    ValNode<V> oldCacheEntry = null;
+    ValNode<V> cacheEntry;
+    while (true) {
+      // repeat until we get a cacheEntry that is not in the process of being removed.
+      cacheEntry = mCache.computeIfAbsent(key, k -> {
+        if (mCache.size() >= mHardLimit) {
+          return null;
+        } else {
+          return new ValNode<>(mDefaultLoader.apply(k));
+        }
+      });
+
+      if (cacheEntry == null) {
+        // cache is at hard limit
+        try {
+          Thread.sleep(100);
+          evictIfOverLimit();
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        }
+        continue;
+      }
+
+      if ((oldCacheEntry != cacheEntry)
+          && cacheEntry.mRefCount.incrementAndGet() > 0) {
+        // refCount went negative, we are evicting this entry
+        cacheEntry.mIsNew = false;
```

Comment:
Could this happen?

1. Thread A creates a cacheEntry (mIsNew = true)
2. Thread B uses the cacheEntry (sets mIsNew = false)
3. Thread B releases the cacheEntry (still mIsNew = false)
4. The cache entry gets evicted
5. Thread A needs to create a new entry

If this can happen even in the presence of `mIsNew`, I don't see much value in maintaining the field

Reply:
maybe `isAccessed` could be a timestamp, and use that info for eviction?

Reply:
this should be fixed by moving incRef into compute